### ISSUE TITLE
Unexpected behavior in comma-dangle, always-multiline: true

### ISSUE THREAD

**Issue by ffxsam** [eslint/eslint#5469]
version 2.2.0

ESLint raised a warning for the following code:

```javascript
const oneThing = [
  2, 8
];
```

IMO, it should not, as the array items are on a single line.

**Comment by platinumazure** [comment]
The expression is multiline, though. An argument could be made either way...

**Comment by eslintbot** [comment]
Hi @ffxsam, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by ffxsam** [comment]
True. But think about the purpose of dangling commas. It's to prevent mistakes where one might have something like this:

```
things = {
  'bread',
  'eggs',
  'cheese'
};
```

And then they add one more item to the list and forget the comma:

```
things = {
  'bread',
  'eggs',
  'cheese'
  'yogurt'
};
```

I think what matters more is, if the list of items are on separate lines, not whether the statement itself is multi-line.

**Comment by platinumazure** [comment]
I guess I'm not sure what the original intent of the rule was supposed to be. This could be a bug, or it might be seen as an enhancement request (possibly for a new option).

@eslint/eslint-team Thoughts?

**Comment by mysticatea** [comment]
Thank you for this issue.

[The document](http://eslint.org/docs/rules/comma-dangle#always-multiline) says

```js
var arr = [1,
    2];
```

is valid. 
I guess the rule is handling a statement as multiline when there is a line break before the close parenthesis. If we change the behavior, it's a breaking change.

Or an option?

**Comment by ffxsam** [comment]
Hmm.. lemme give it some careful consideration and I’ll post a proposal in this thread.

**Comment by nzakas** [comment]
>  If we change the behavior, it's a breaking change.

Not necessarily, I think this looks more like a bug. Multiline is meant for items, not the brackets.

**Comment by alberto** [comment]
That's debatable. From the rule docs:
> trailing commas simplify adding and removing items to objects and arrays, since only the lines you are modifying must be touched.

**Comment by nzakas** [comment]
@alberto that quote doesn't seem to counter what I said.

**Comment by alberto** [comment]
I think it is also desirable to allow going from this:
```
const foo = [
  1,
];
```
to this:
```
const foo = [
  1,
  2,
];
```
without having to add the comma on the previous line.

**Comment by ffxsam** [comment]
My vote is still to make `always-multiline: true` not complain about this:

```
const oneThing = [
  2, 8
];
```

since the actual items are on a single line. IMO it should really be about the _actual_ items, not the surrounding syntax.

**Comment by mysticatea** [comment]
So multiline is:

- If the number of items is `0` or `1`, when there is a line break before the close paren (the current behavior).
- Otherwise, when the last item exists at a different line with the line of the first item.

?

**Comment by ffxsam** [comment]
I think that sounds right. Just to clarify:

For the rule `"comma-dangle": [2, "always-multiline"]` to validate, dangling commas must be used in the following situation:

* When the last item in a list is on a line by itself

I think it really boils down to that one simple rule. So, with `always-multiline` enabled, all of these will throw warnings/errors:

```
const x = [
  2
];

const x = [
  3,
  8
];

const x = [
  {
    item: "rock",
    weight: 8,
  },
  {
    item: "shovel",
    weight: 14,
  }
];
```

While these are totally valid:

```
const x = [23];

const x = [
  4,
  20,
];

const x = [{name: 'Jean', age: 44}, {name: 'Remy', age: 31}];
```

The reasoning behind this logic is that in a list like this:

```
[
  'hotdog',
  'hamburger'
]
```

it's likely that you'll want to add onto it. And if you were to add a third item, it would result in *two* lines being modified, one to add a comma after hamburger, and one for the new item. This is actually another reason for dangling commas, to reduce the number of git diff lines.

Whereas in this list:

```
['hotdog', 'hamburger']
```

The dangling comma doesn't matter, because adding on a third item will change just a single line whether there's a dangling comma or not.

So I think it comes down to the determining factor being whether the last item is on a line by itself or not. Thoughts?

**Comment by nzakas** [comment]
@mysticatea :+1:

**Comment by BYK** [comment]
@mysticatea - are you working on this?

**Comment by mysticatea** [comment]
No, I don't have enough time now. :cry:

**Comment by alberto** [comment]
I'll take it.

**Comment by BenoitZugmeyer** [comment]
I may be a little late to argue against this change, but in my mind the `always-multiline` aims to enforce clean (line) diff when adding or removing items in an array or object. Thus, optional comma at the end of any line voids the purpose of this rule.

**Comment by nzakas** [comment]
Can everyone please review the new behavior in the pull request? 
https://github.com/eslint/eslint/pull/5570

@BenoitZugmeyer it's not clear from your description what exactly you're saying. Can you show some code examples that explain your case?

**Comment by BenoitZugmeyer** [comment]
@nzakas if you have an array declaration like this:
```
var tags = [
  'div', 'img', ..., 'ul',
  'ul', 'li', 'a', ..., 'span'
]
```
and you want to add a new tag name on a new line, you will have to add a comma at the end of the last existing line:
```
var tags = [
  'div', 'img', ..., 'ul',
  'ul', 'li', 'a', ..., 'span',
  'video'
]
```
The diff will be over two lines instead of only one if the comma was already here. On top of this, it can be the source of merge conflicts if two different commits add the comma.

This may be a small issue but I use this option mostly to avoid this.

**Comment by alberto** [comment]
@BenoitZugmeyer that will still be considered a warning

**Comment by nzakas** [comment]
@BenoitZugmeyer sorry, I'm not following. What are you expecting to be flagged? What would be different with the new implementation?

**Comment by BenoitZugmeyer** [comment]
Sorry if I'm not clear enough. With the current implementation, the following snippet will raise a warning (as I expect), because the array declaration is on multiple lines and there is no comma after `2`:
```
a = [
  1, 2
]
```
With the new implementation, it won't raise any warning: the comma after `2` is optional. This is the whole purpose of this issue right?

On top of being less strict about code styling, I like the current behavior because it enforces clean diffs. Example, if I want to add 3 and 4 on a new line with the previous snippet, the diff would look like

```diff
 a = [
-  1, 2
+  1, 2,
+  3, 4
]
```

See? Two added lines. With the current implementation, enforcing a comma after `2`, the comma would already be here, and only one line would be added in the diff.

**Comment by ffxsam** [comment]
You're right. I didn't think of that scenario because I actually never lay out my lists in a matrix type format like that. But it's a valid point.

**Comment by nzakas** [comment]
@BenoitZugmeyer thanks, that helps. 

So maybe instead of making a change to default behavior, we should add an option to toggle it?

**Comment by alberto** [comment]
@BenoitZugmeyer yes, if all the elements are on the same line, (unlike in your previous examples) it won't cause an error.

**Comment by BenoitZugmeyer** [comment]
@nzakas maybe a separate flag, like `"always-multirow"` meaning the structure contains multiple rows (versus spans on multiple lines)

@alberto thanks for noticing, I edited my previous example to fix this

**Comment by alberto** [comment]
So, should this go behind a flag, then?

**Comment by nzakas** [comment]
I think so. @BenoitZugmeyer made a pretty strong case.

**Comment by nzakas** [comment]
That probably means this is also an enhancement now instead of a bug.

**Comment by alberto** [comment]
Is the proposed name `always-multirow` ok for you?

This goes back to whether multiline is meant for items or brackets. @BenoitZugmeyer considers them, @ffxsam doesn't.

So maybe something like `ignoreBrackets` or `ignoreBracketsPosition`?

**Comment by ffxsam** [comment]
No, I'm on board with @BenoitZugmeyer now.

**Comment by alberto** [comment]
Should we close this then?

**Comment by ffxsam** [comment]
I'm cool with that if everyone else is.

**Comment by alberto** [comment]
@nzakas what do you want to do?

**Comment by nzakas** [comment]
The best work is no work. Let's close it.
